doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+
+#### `SSL_tick`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | Yes       | CS            |
+
+Advances the QUIC state machine to the extent feasible, potentially performing
+network I/O. Also compatible with DTLSv1 and supercedes `DTLSv1_handle_timeout`
+for all use cases.
+
+TBD: Should we just map this to DTLS_CTRL_HANDLE_TIMEOUT internally (and maybe
+alias the CTRL #define)?
+
+#### `SSL_get_tick_timeout`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+Gets the time until the QUIC state machine next wants to receive a timeout
+event, if any.
+
+This is similar to the existing `DTLSv1_get_timeout` function, but it is not
+specific to DTLSv1. It is also usable for DTLSv1 and can become a
+protocol-agnostic API for this purpose, superceding `DTLSv1_get_timeout` for all
+use cases.
+
+The design is similar to that of `DTLSv1_get_timeout` and uses a `struct
+timeval`. However, this function represents an infinite timeout (i.e., no
+timeout) using `tv_sec == -1`, whereas `DTLSv1_get_timeout` represents an
+infinite timeout using a 0 return value, which does not allow a failure
+condition to be distinguished.
+
+TBD: Should we just map this to DTLS_CTRL_GET_TIMEOUT internally (and maybe
+alias the CTRL #define)?
+
+#### `SSL_set_blocking_mode`, `SSL_get_blocking_mode`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+Turns blocking mode on or off. This is necessary because up until now libssl has
+operated in blocking or non-blocking mode automatically as an emergent
+consequence of whether the underlying network socket is blocking. Since we are
+proposing to use only non-blocking I/O internally, use of blocking semantics at
+the application level must be explicitly configured.
+
+Use on stream objects: It may be feasible to implement this such that different
+QUIC stream SSL objects can have different settings for this option.
+
+Not supported for non-QUIC SSL objects.
+
+#### `SSL_get_rpoll_descriptor`, `SSL_get_wpoll_descriptor`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+These functions output poll descriptors which can be used to determine
+when the QUIC state machine should next be ticked. `SSL_get_rpoll_descriptor` is
+relevant if `SSL_want_net_read` returns 1, and `SSL_get_wpoll_descriptor` is
+relevant if `SSL_want_net_write` returns 1.
+
+The implementation of these functions is a simple forward to
+`BIO_get_rpoll_descriptor` and `BIO_get_wpoll_descriptor` on the underlying
+network BIOs.
```

Comment:
What happens for non-QUIC objects?

Reply:
I see no reason these can't be implemented on non-QUIC objects too.